sklearn/manifold/t_sne.py

```diff
@@ -753,34 +756,40 @@ def _fit(self, X, skip_num_points=0):
         if self.method == 'barnes_hut':
             if self.verbose:
                 print("[t-SNE] Computing %i nearest neighbors..." % k)
+
+            # Find the nearest neighbors for every point
+            # TODO: argument for class knn_estimator=None
+            # TODO: assert that the knn metric is euclidean
             if self.metric == 'precomputed':
-                # Use the precomputed distances to find
-                # the k nearest neighbors and their distances
-                neighbors_nn = np.argsort(distances, axis=1)[:, :k]
+                knn = NearestNeighbors(metric=self.metric)
             else:
-                # Find the nearest neighbors for every point
-                bt = BallTree(X)
-                # LvdM uses 3 * perplexity as the number of neighbors
-                # And we add one to not count the data point itself
-                # In the event that we have very small # of points
-                # set the neighbors to n - 1
-                distances_nn, neighbors_nn = bt.query(X, k=k + 1)
-                neighbors_nn = neighbors_nn[:, 1:]
-            P = _joint_probabilities_nn(distances, neighbors_nn,
+                knn = NearestNeighbors(algorithm='ball_tree',
+                                       metric=self.metric)
+            knn.fit(X)
+            # LvdM uses 3 * perplexity as the number of neighbors
+            # And we add one to not count the data point itself
+            # In the event that we have very small # of points
+            # set the neighbors to n - 1
+            distances_nn, neighbors_nn = knn.kneighbors(
```

Comment:
We don't need any of the feature of the scipy sparse matrix API for the neighbors info itself (we do Cython code directly).

Only the conditional probability matrix benefits from being represented as a scipy sparse matrix to make it a one liner python snippet to do the symmetrization.